From ebc1e4a8fbbe05eda036cf435ef701d3ee27995d Mon Sep 17 00:00:00 2001
From: Lin Fuzhen <fuzhen.lin@freescale.com>
Date: Thu, 25 Apr 2013 18:05:39 +0800
Subject: [PATCH 195/602] ENGR00260448 i.MX5x/i.MX6x: Add RecoveryUI for i.MX
 SoCs

Add RecoveryUI for i.MX5x/6x reference boards which at least have three
phycial keys: power key, volume up and volume down.

In recovery mode, press and hold power key, then press the volume up key
will toggle the recovery menu and entry in Text recovery mode.
In test recovery mode, using volume up/down key to select the action,and
press power key to invloke the select item.

Signed-off-by: Lin Fuzhen <fuzhen.lin@freescale.com>
---
 common/recovery/Android.mk      | 13 ++++++
 common/recovery/recovery_ui.cpp | 95 +++++++++++++++++++++++++++++++++++++++++
 imx53_smd/BoardConfig.mk        |  1 -
 imx5x/BoardConfigCommon.mk      |  1 +
 imx6/BoardConfigCommon.mk       |  1 +
 5 files changed, 110 insertions(+), 1 deletion(-)
 create mode 100644 common/recovery/Android.mk
 create mode 100644 common/recovery/recovery_ui.cpp

diff --git a/common/recovery/Android.mk b/common/recovery/Android.mk
new file mode 100644
index 0000000..f98468b
--- /dev/null
+++ b/common/recovery/Android.mk
@@ -0,0 +1,13 @@
+LOCAL_PATH := $(call my-dir)
+include $(CLEAR_VARS)
+
+LOCAL_MODULE_TAGS := eng
+LOCAL_C_INCLUDES += bootable/recovery
+LOCAL_SRC_FILES := recovery_ui.cpp
+
+# should match TARGET_RECOVERY_UI_LIB set in BoardConfigCommon.mk
+LOCAL_MODULE := librecovery_ui_imx
+
+include $(BUILD_STATIC_LIBRARY)
+
+include $(CLEAR_VARS)
diff --git a/common/recovery/recovery_ui.cpp b/common/recovery/recovery_ui.cpp
new file mode 100644
index 0000000..ccf8ccd
--- /dev/null
+++ b/common/recovery/recovery_ui.cpp
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2011 The Android Open Source Project
+ * Copyright (C) 2013 Freescale Semiconductor, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <linux/input.h>
+#include <sys/stat.h>
+#include <errno.h>
+#include <string.h>
+
+#include "common.h"
+#include "device.h"
+#include "screen_ui.h"
+
+const char* HEADERS[] = { "Volume up/down to move highlight;",
+                          "power button to select.",
+                          "",
+                          NULL };
+
+const char* ITEMS[] = { "reboot system now",
+                        "apply update from ADB",
+                        "wipe data/factory reset",
+                        "wipe cache partition",
+                        NULL };
+
+class ImxUI : public ScreenRecoveryUI {
+  public:
+    virtual KeyAction CheckKey(int key) {
+        if (IsKeyPressed(KEY_POWER) && key == KEY_VOLUMEUP) {
+            return TOGGLE;
+        }
+        return ENQUEUE;
+    }
+};
+
+
+class ImxDevice : public Device {
+  public:
+    ImxDevice() :
+        ui(new ImxUI) {
+    }
+
+    RecoveryUI* GetUI() { return ui; }
+
+    int HandleMenuKey(int key_code, int visible) {
+        if (visible) {
+            switch (key_code) {
+              case KEY_DOWN:
+              case KEY_VOLUMEDOWN:
+                return kHighlightDown;
+
+              case KEY_UP:
+              case KEY_VOLUMEUP:
+                return kHighlightUp;
+
+              case KEY_POWER:
+                return kInvokeItem;
+            }
+        }
+
+        return kNoAction;
+    }
+
+    BuiltinAction InvokeMenuItem(int menu_position) {
+        switch (menu_position) {
+          case 0: return REBOOT;
+          case 1: return APPLY_ADB_SIDELOAD;
+          case 2: return WIPE_DATA;
+          case 3: return WIPE_CACHE;
+          default: return NO_ACTION;
+        }
+    }
+
+    const char* const* GetMenuHeaders() { return HEADERS; }
+    const char* const* GetMenuItems() { return ITEMS; }
+
+  private:
+    RecoveryUI* ui;
+};
+
+Device* make_device() {
+    return new ImxDevice;
+}
diff --git a/imx53_smd/BoardConfig.mk b/imx53_smd/BoardConfig.mk
index b114b15..100520b 100755
--- a/imx53_smd/BoardConfig.mk
+++ b/imx53_smd/BoardConfig.mk
@@ -76,7 +76,6 @@ BOARD_KERNEL_BASE := 0x70800000
 
 TARGET_BOOTLOADER_CONFIG := mx53_smd_android_config
 
-TARGET_RECOVERY_UI_LIB := librecovery_ui_imx53_smd
 
 # define frame buffer count
 NUM_FRAMEBUFFER_SURFACE_BUFFERS := 3
diff --git a/imx5x/BoardConfigCommon.mk b/imx5x/BoardConfigCommon.mk
index 45c5368..5c4c230 100755
--- a/imx5x/BoardConfigCommon.mk
+++ b/imx5x/BoardConfigCommon.mk
@@ -68,3 +68,4 @@ TARGET_PROVIDES_INIT_RC := true
 
 BOARD_SYSTEMIMAGE_PARTITION_SIZE := 293601280
 BOARD_FLASH_BLOCK_SIZE := 4096
+TARGET_RECOVERY_UI_LIB := librecovery_ui_imx
diff --git a/imx6/BoardConfigCommon.mk b/imx6/BoardConfigCommon.mk
index 81325a9..5125b0a 100644
--- a/imx6/BoardConfigCommon.mk
+++ b/imx6/BoardConfigCommon.mk
@@ -59,5 +59,6 @@ BOARD_RECOVERY_PARTITION_SIZE = 10M
 
 BOARD_SYSTEMIMAGE_PARTITION_SIZE := 293601280
 BOARD_FLASH_BLOCK_SIZE := 4096
+TARGET_RECOVERY_UI_LIB := librecovery_ui_imx
 
 -include device/google/gapps/gapps_config.mk
-- 
1.8.0

